java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java

```diff
@@ -156,7 +126,55 @@ public String toString() {
 
     @CheckForNull
     public MethodYield selectedMethodYield(Node from) {
-      return parents.get(from);
+      return edges.containsKey(from) ? edges.get(from).yields.stream().findFirst().orElse(null) : null;
+    }
+  }
+
+  static final class Edge {
+    final Node child;
+    final Node parent;
+
+    private Set<LearnedConstraint> lc;
+    private Set<LearnedAssociation> la;
+    private final Set<MethodYield> yields = new LinkedHashSet<>();
+
+    @VisibleForTesting
+    Edge(Node child, Node parent) {
+      Preconditions.checkState(!Objects.equals(child, parent));
+      this.child = child;
+      this.parent = parent;
+    }
+
+    Set<LearnedConstraint> learnedConstraints() {
+      if (lc == null) {
+        lc = child.programState.learnedConstraints(parent.programState);
+      }
+      return lc;
+    }
+
+    Set<LearnedAssociation> learnedAssociations() {
+      if (la == null) {
+        la = child.programState.learnedAssociations(parent.programState);
+      }
+      return la;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
unsure you actually need to implement hashcode/equals for edges. IMO those should be dropped for the time being.